disnake/ext/commands/params.py

```diff
@@ -403,97 +473,220 @@ def to_option(self) -> Option:
+                args = args[:index]
+                kwargs = {}
+            break
+        if posarg is None:
+            parsed_pos = True
+        if parsed_pos and parameter.name in possible_kwargs:
+            kwargs[parameter.name] = possible_kwargs[parameter.name]
+
+    return function(*args, **kwargs)
+
+
+def isolate_self(
+    function: Callable,
+) -> Tuple[Tuple[Optional[inspect.Parameter], ...], Dict[str, inspect.Parameter]]:
+    """Create parameters without self and the first interaction"""
+    is_interaction = (
+        lambda annot: issubclass_(annot, Interaction) or annot is inspect.Parameter.empty
     )
-    inter_param = disnake.utils.find(predicate, parameters)
-    parameters = parameters[parameters.index(inter_param) + 1 :]
-    type_hints = get_type_hints(command.callback)
-    docstring = command.docstring["params"]
-
-    # extract params:
-    params = []
-    for parameter in parameters:
-        if parameter.kind in [inspect.Parameter.VAR_POSITIONAL, inspect.Parameter.VAR_KEYWORD]:
-            continue
-        param = ParamInfo.from_param(parameter, type_hints, docstring)
-        params.append(param)
 
-    # update connectors and autocompleters
-    for param in params:
-        if param.name != param.param_name:
-            command.connectors[param.name] = param.param_name
-        if param.autocomplete:
-            command.autocompleters[param.name] = param.autocomplete
+    sig = signature(function)
 
-    # add custom decorators
-    inter_annot = type_hints.get(inter_param.name, Any)
-    if isinstance(inter_annot, type) and issubclass(inter_annot, disnake.GuildCommandInteraction):
-        command.guild_only = True
+    parameters = dict(sig.parameters)
+    parametersl = list(sig.parameters.values())
 
-    return [param.to_option() for param in params]
+    if not parameters:
+        return (None, None), {}
 
+    cog_param: Optional[inspect.Parameter] = None
+    inter_param: Optional[inspect.Parameter] = None
 
-async def resolve_param_kwargs(
-    func: Callable, inter: Interaction, kwargs: Dict[str, Any]
-) -> Dict[str, Any]:
-    """Resolves a call with kwargs and transforms into normal kwargs
+    if parametersl[0].name == "self":
+        cog_param = parameters.pop(parametersl[0].name)
+        parametersl.pop(0)
+        if len(parameters) > 1 and is_interaction(parametersl[0].annotation):
+            inter_param = parameters.pop(parametersl[0].name)
+    if inter_param is None and is_interaction(parametersl[0].annotation):
+        inter_param = parameters.pop(parametersl[0].name)
 
-    Depends on the fact that optionparams already contain all info.
+    return (cog_param, inter_param), parameters
+
+
+def collect_params(
+    function: Callable,
+) -> Tuple[Optional[str], Optional[str], List[ParamInfo], Dict[str, Injection]]:
+    """Collect all parameters in a function
+
+    Returns: (`cpg parameter`, `interaction parameter`, `param infos`, `injections`)
```

Comment:
```suggestion
    Returns: (`cog parameter`, `interaction parameter`, `param infos`, `injections`)
```
I know this is just docstrings.